library/core/src/pin.rs

```diff
@@ -1,188 +1,621 @@
+//! move a `T` out of a [`&mut T`]. Therefore, putting a value (such as `tracker` above) behind a
+//! pointer isn't enough on its own to ensure that its address does not change.
+//!
+//! # What is "pinning"?
+//! [what-is-pinning]: self#what-is-pinning
+//!
+//! We say that a value has been *pinned* when it has been put into a state where it is guaranteed
+//! to remain *valid* and *located at the same place in memory* from the time it is pinned until its
+//! [`drop`] is called.
+//!
+//! ## Address-sensitive values, AKA "when we need pinning"
+//! [address-sensitive-values]: self#address-sensitive-values-aka-when-we-need-pinning
+//!
+//! Most values in Rust are entirely okay with being *moved* around at-will.
+//! Types for which it is *always* the case that *any* value of that type can be
+//! *moved* at-will should implement [`Unpin`], which we will discuss more [below][self#unpin].
+//!
+//! [`Pin`] is specifically targeted at allowing the implementation of *safe interfaces* around
+//! types which have some state during which they become "address-sensitive." A value in such an
+//! "address-sensitive" state is *not* okay with being *moved* around at-will. Such a value must
+//! stay *un-moved* and valid during the address-sensitive portion of its lifetime because some
+//! interface is relying on those invariants to be true in order for its implementation to be sound.
+//!
+//! As a motivating example of a type which may become address-sensitive, consider a type which
+//! contains a pointer to another piece of its own data, *i.e.* a "self-referential" type. In order
+//! for such a type to be implemented soundly, the pointer which points into `self`'s data must be
+//! proven valid whenever it is accessed. But if that value is *moved*, the pointer will still
+//! point to the old address where the value was located and not into the new location of `self`,
+//! thus becoming invalid. A key example of such self-referrential types are the state machines
+//! generated by the compiler to implement [`Future`] for `async fn`s.
+//!
+//! Such types that have an *address-sensitive* state usually follow a lifecycle
+//! that looks something like so:
+//!
+//! 1. A value is created which can be freely moved around.
+//!     * e.g. calling an async function which returns a state machine implementing [`Future`]
+//! 2. An operation causes the value to depend on its own address not changing
+//!     * e.g. calling [`poll`] for the first time on the produced [`Future`]
+//! 3. Further pieces of the safe interface of the type use internal [`unsafe`] operations which
+//! assume that the address of the value is stable
+//!     * e.g. subsequent calls to [`poll`]
+//! 4. Before the value is invalidated (e.g. deallocated), it is *dropped*, giving it a chance to
+//! notify anything with pointers to itself that those pointers will be invalidated
+//!     * e.g. [`drop`]ping the [`Future`]
```

Comment:
This is stretching it a bit since futures never care about the drop guarantee. Only intrusive collections do.

Not sure what to do about that, just pointing this out as a potential concern/confusion.

Reply:
sounds like something that can go in a footnote